runtime/compiler/control/JITServerHelpers.cpp

```diff
@@ -486,18 +486,32 @@ JITServerHelpers::printJITServerCacheStats(J9JITConfig *jitConfig, TR::Compilati
       }
    }
 
-bool 
-JITServerHelpers::cacheRemoteROMClass(ClientSessionData *clientSessionData, J9Class *clazz, J9ROMClass *romClass, ClassInfoTuple *classInfoTuple)
+/*
+  `cacheRemoteROMClassOrFreeIt` takes a romClass data structure allocated with
+  persistent memory and attempts to cache it in the per-client data session. 
+  If caching suceeds, the method returns a pointer to the romClass received as parameter. 
+  If the caching fails, the memory for romClass received as parameter is freed 
+  and the method returns a pointer to the romClass from the cache
+ */
+J9ROMClass *
+JITServerHelpers::cacheRemoteROMClassOrFreeIt(ClientSessionData *clientSessionData, J9Class *clazz, J9ROMClass *romClass,
+                                              ClassInfoTuple *classInfoTuple, TR_PersistentMemory *persistentMemory)
    {
    ClientSessionData::ClassInfo classInfo;
    OMR::CriticalSection cacheRemoteROMClass(clientSessionData->getROMMapMonitor());
    auto it = clientSessionData->getROMClassMap().find((J9Class*)clazz);
    if (it == clientSessionData->getROMClassMap().end())
       {
       JITServerHelpers::cacheRemoteROMClass(clientSessionData, clazz, romClass, classInfoTuple, classInfo);
-      return true;
-      }
-   return false;
+      return romClass;
+      }
+   // romClass is already present in the cache; must free the duplicate
+   if (auto cache = TR::CompilationInfo::get()->getJITServerSharedROMClassCache()) 
```

Comment:
I think it would be better to put this in a separate helper method (e.g. `freeRemoteROMClass()`) and also call it in `ClientSessionData::ClassInfo::freeClassInfo()` to avoid code duplication

Reply:
Done.